modules/space/horizonsfile.cpp

```diff
@@ -0,0 +1,877 @@
+            );
+        }
+    }
+    else {
+        LWARNING("Could not find signature information");
+    }
+
+    // Errors
+    if (auto it = answer.find("error"); it != answer.end()) {
+        // There was an error
+        std::string errorMsg = *it;
+
+        // Projected output length (~X) exceeds 90024 line max -- change step-size
+        if (errorMsg.find("Projected output length") != std::string::npos) {
+            return ResultCode::ErrorSize;
+        }
+        // STEP_SIZE too big, exceeds available span.
+        else if (errorMsg.find("STEP_SIZE too big") != std::string::npos) {
+            return ResultCode::ErrorSpan;
+        }
+        // No ephemeris for target "X" after A.D. Y UT
+        else if (errorMsg.find("No ephemeris for target") != std::string::npos) {
+            return ResultCode::ErrorTimeRange;
+        }
+        // No site matches. Use "*@body" to list, "c@body" to enter coords, ?! for help.
+        else if (errorMsg.find("No site matches") != std::string::npos ||
+                 errorMsg.find("Cannot find central body") != std::string::npos)
+        {
+            return ResultCode::ErrorNoObserver;
+        }
+        // Observer table for X / Y->Y disallowed.
+        else if (errorMsg.find("disallowed") != std::string::npos) {
+            return ResultCode::ErrorObserverTargetSame;
+        }
+        // Insufficient ephemeris data has been loaded to compute the state of X
+        // relative to Y at the ephemeris epoch Z;
+        else if (errorMsg.find("Insufficient ephemeris data") != std::string::npos) {
+            return ResultCode::ErrorNoData;
+        }
+        // #   E. Lon    DXY      DZ    Observatory Name;
+        // -- - -------- ------ - ------ - ----------------;
+        // * Observer station *
+        // Multiple matching stations found.
+        else if (errorMsg.find("Multiple matching stations found") != std::string::npos) {
+            return ResultCode::MultipleObserverStations;
+        }
+        // Unknown error
+        else {
+            LERROR(errorMsg);
+            return ResultCode::UnknownError;
+        }
+    }
+    return ResultCode::Valid;
+}
+
+// Check whether the given Horizons file is valid or not
+// Return an error code with what is the problem if there was one
+HorizonsFile::ResultCode HorizonsFile::isValidHorizonsFile(std::filesystem::path file) {
+    std::ifstream fileStream(file);
+    if (!fileStream.good()) {
+        return ResultCode::Empty;
+    }
+
+    // The header of a Horizons file has a lot of information about the
+    // query that can tell us if the file is valid or not.
+    // The line $$SOE indicates start of data.
+    std::string line;
+    bool foundTarget = false;
+    std::getline(fileStream, line);
+    std::getline(fileStream, line); // First line is just stars (*) no information, skip
+
+    // Valid Target?
+    if (fileStream.good() &&
+        (line.find("Revised") != std::string::npos ||
```

Comment:
Misaligned if statement (and maybe also a place where we could use a helper function to clean up the check if string contains a substring?)